File: frontend/src/api/historical.ts
import { api } from "./client";

export interface DataPoint {
  datetime: string;
  demand_mw: number;
}

export interface HistoricalResponse {
  data: DataPoint[];
  total_records: number;
  aggregation: string;
}

export async function fetchHistorical(
  startDate?: string,
  endDate?: string,
  aggregation: string = "daily"
): Promise<HistoricalResponse> {
  const params: Record<string, string> = { aggregation };
  if (startDate) params.start_date = startDate;
  if (endDate) params.end_date = endDate;
  const { data } = await api.get<HistoricalResponse>("/historical", { params });
  return data;
}

File: backend/app/routers/historical.py
from fastapi import APIRouter, Request, Query
from app.schemas.historical import HistoricalResponse
from app.services.data_service import filter_data, aggregate_data

router = APIRouter()


@router.get("/historical", response_model=HistoricalResponse)
def get_historical(
    request: Request,
    start_date: str | None = Query(None),
    end_date: str | None = Query(None),
    aggregation: str = Query("daily", regex="^(hourly|daily|weekly|monthly)$"),
):
    df = request.app.state.data.copy()
    df = filter_data(df, start_date, end_date)
    data = aggregate_data(df, aggregation)
    return HistoricalResponse(
        data=data,
        total_records=len(data),
        aggregation=aggregation,
    )
